### PR TITLE
Tiger implemented

### LINKED ISSUE #49: Implement Tiger hash function

### PR DESCRIPTION
**Pull request by nkaskov**
Closes #49

### CODE REVIEW

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                    typedef std::array<word_type, block_words> block_type;
+
+                    constexpr static const std::size_t state_bits = DigestBits;
+                    constexpr static const std::size_t state_words = state_bits / word_bits;
+                    typedef std::array<word_type, state_words> state_type;
+
+                    constexpr static const std::size_t substitution_words = 256;
+                    typedef std::array<word_type, substitution_words> substitution_type;
+
+                    constexpr static const substitution_type sbox1 = {
```
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:37]: Just thinking out loud here. Has anyone considered making this compatible (or just testing, if it is compatible) with C++ standards lower, than C++17? I mean ```constexpr inline``` variables, as they are being used in here, were introduced only in C++17, as far as I remember (https://en.cppreference.com/w/cpp/language/inline). Earlier standards require different specifications. We have ```CONSTEXPR_INLINE_VARIABLE``` for this.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                        0x67E366B35C3C432C, 0xB923DD61102B37F2, 0x56AB2779D884271D, 0xBE83E1B0FF1525AF,
+                        0xFB7C65D4217E49A9, 0x6BDBE0E76D48E7D4, 0x08DF828745D9179E, 0x22EA6A9ADD53BD34,
+                        0xE36E141C5622200A, 0x7F805D1B8CB750EE, 0xAFE5C7A59F58E837, 0xE27F996A4FB1C23C,
+                        0xD3867DFB0775F0D0, 0xD0E673DE6E88891A, 0x123AEB9EAFB86C25, 0x30F1D5D5C145B895,
+                        0xBB434A2DEE7269E7, 0x78CB67ECF931FA38, 0xF33B0372323BBF9C, 0x52D66336FB279C74,
+                        0x505F33AC0AFB4EAA, 0xE8A5CD99A2CCE187, 0x534974801E2D30BB, 0x8D2D5711D5876D90,
+                        0x1F1A412891BC038E, 0xD6E2E71D82E56648, 0x74036C3A497732B7, 0x89B67ED96361F5AB,
+                        0xFFED95D8F1EA02A2, 0xE72B3BD61464D43D, 0xA6300F170BDC4820, 0xEBC18760ED78A77A};
+
+                    constexpr static const substitution_type sbox2 = {
```
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:103]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                        0x6B4CB165F9CF0E5A, 0x43AA1C06A0ABFB4A, 0x7F4DC26FF162796B, 0x6CBACC8E54ED9B0F,
+                        0xA6B7FFEFD2BB253E, 0x2E25BC95B0A29D4F, 0x86D6A58BDEF1388C, 0xDED74AC576B6F054,
+                        0x8030BDBC2B45805D, 0x3C81AF70E94D9289, 0x3EFF6DDA9E3100DB, 0xB38DC39FDFCC8847,
+                        0x123885528D17B87E, 0xF2DA0ED240B1B642, 0x44CEFADCD54BF9A9, 0x1312200E433C7EE6,
+                        0x9FFCC84F3A78C748, 0xF0CD1F72248576BB, 0xEC6974053638CFE4, 0x2BA7B67C0CEC4E4C,
+                        0xAC2F4DF3E5CE32ED, 0xCB33D14326EA4C11, 0xA4E9044CC77E58BC, 0x5F513293D934FCEF,
+                        0x5DC9645506E55444, 0x50DE418F317DE40A, 0x388CB31A69DDE259, 0x2DB4A83455820A86,
+                        0x9010A91E84711AE9, 0x4DF7F0B7B1498371, 0xD62A2EABC0977179, 0x22FAC097AA8D5C0E};
+
+                    constexpr static const substitution_type sbox3 = {
```
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:169]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                        0xE10FDD37F7882A90, 0x524DCAB5518A3F5C, 0x3C9E85878451255B, 0x4029828119BD34E2,
+                        0x74A05B6F5D3CECCB, 0xB610021542E13ECA, 0x0FF979D12F59E2AC, 0x6037DA27E4F9CC50, 
+                        0x5E92975A0DF1847D, 0xD66DE190D3E623FE, 0x5032D6B87B568048, 0x9A36B7CE8235216E, 
+                        0x80272A7A24F64B4A, 0x93EFED8B8C6916F7, 0x37DDBFF44CCE1555, 0x4B95DB5D4B99BD25,
+                        0x92D3FDA169812FC0, 0xFB1A4A9A90660BB6, 0x730C196946A4B9B2, 0x81E289AA7F49DA68, 
+                        0x64669A0F83B1A05F, 0x27B3FF7D9644F48B, 0xCC6B615C8DB675B3, 0x674F20B9BCEBBE95, 
+                        0x6F31238275655982, 0x5AE488713E45CF05, 0xBF619F9954C21157, 0xEABAC46040A8EAE9,
+                        0x454C6FE9F2C0C1CD, 0x419CF6496412691C, 0xD3DC3BEF265B0F70, 0x6D0E60F5C3578A9E };
+
+                    constexpr static const substitution_type sbox4 = {
```
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:235]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp (hunk 1 above)
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:37]: Just thinking out loud here. Has anyone considered making this compatible (or just testing, if it is compatible) with C++ standards lower, than C++17? I mean ```constexpr inline``` variables, as they are being used in here, were introduced only in C++17, as far as I remember (https://en.cppreference.com/w/cpp/language/inline). Earlier standards require different specifications. We have ```CONSTEXPR_INLINE_VARIABLE``` macro for this.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp (hunk 2 above)
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:103]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp (hunk 3 above)
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:169]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp (hunk 4 above)
Comment (nemothenoone) [inline include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:235]: Same.

File: include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp (hunk 1 above)
Comment (nkaskov) [reply include/nil/crypto3/hash/detail/tiger/basic_tiger_policy.hpp:37]: According to the results of Travis CI build, current implementation works for old compiler versions.

**Pull request merged**